### PR TITLE
Precompile TensorProductMatrixSymmetricSum kernels

### PR DESCRIPTION
**Pull request by peterrum**

### CODE REVIEW

File: include/deal.II/lac/tensor_product_matrix.templates.h [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/lac/tensor_product_matrix.h>
+
+#ifndef FDM_DEGREE_MAX
+// We set this value 17, since this is the value needed for smoothers for
+// cell-centered patches with overlap for constinuous elements with degrees up
+// to FE_EVAL_FACTORY_DEGREE_MAX.
+#  define FDM_DEGREE_MAX 17
```
Comment (peterrum) [inline include/deal.II/lac/tensor_product_matrix.templates.h:28]: This should be coupled to 
https://github.com/dealii/dealii/blob/dfde63a25cf241f8ef3e67f8e247258cd84de634/include/deal.II/matrix_free/evaluation_template_factory_internal.h#L24

Related to https://github.com/dealii/dealii/issues/14106.

File: include/deal.II/lac/tensor_product_matrix.h [hunk 2]
```diff
@@ -512,61 +674,17 @@ TensorProductMatrixSymmetricSumBase<dim, Number, n_rows_1d>::vmult(
...
-      eval.template apply<0, false, false>(A0, src, t + n);
-      eval.template apply<1, false, true>(M1, t + n, t);
-      eval.template apply<2, false, true>(M2, t, dst);
-    }
+  Number *      dst = dst_view.begin();
+  const Number *src = src_view.begin();
 
+  if (n_rows_1d != -1)
+    internal::TensorProductMatrixSymmetricSum::vmult<
+      n_rows_1d == -1 ? 0 : n_rows_1d>(
```
Comment (kronbichler) [inline include/deal.II/lac/tensor_product_matrix.h:683]: It looks like you will never reach the `n_rows_1d == -1` in this branch, as you check `n_rows_1d != -1` in line 681. Is the code there to reconcile the compiler?

File: include/deal.II/lac/tensor_product_matrix.templates.h [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+#define dealii_tensor_product_matrix_templates_h
+
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/lac/tensor_product_matrix.h>
+
+#ifndef FDM_DEGREE_MAX
+// We set this value 17, since this is the value needed for smoothers for
+// cell-centered patches with overlap for constinuous elements with degrees up
```
Comment (kronbichler) [inline include/deal.II/lac/tensor_product_matrix.templates.h:26]:
```suggestion
// cell-centered patches with overlap for continuous elements with degrees up
```

**Review by kronbichler — APPROVED** [review]
Do we have a test for the two paths, one including argument `-1`, already?

File: include/deal.II/lac/tensor_product_matrix.h (hunk 2 above)
Comment (peterrum) [reply include/deal.II/lac/tensor_product_matrix.h:683]: The line is never reached; but without `if consexpr`, the compiler will still complain.

**Comment by peterrum** [comment]
> Do we have a test for the two paths, one including argument -1, already?

Yes, there are tests!

File: include/deal.II/lac/tensor_product_matrix.h (hunk 2 above)
Comment (kronbichler) [reply include/deal.II/lac/tensor_product_matrix.h:683]: I know, I've had similar constructions before.

**Review by kronbichler — APPROVED** [review]

**Pull request merged**